add basic npm module

zeke opened the pull request:
This PR is an experiment to create an npm package that can be easily consumed by projects like [nodejs/nodejs.dev](https://github.com/nodejs/nodejs.dev), [nodejs/nodejs.org](https://github.com/nodejs/nodejs.org), or others. This pull request is intended to start a discussion about what an ideal module API might look like, with the primary consumer being the new nodejs.dev website.

### Installation

This thing is not yet published to npm, but you can install it directly from this GitHub branch using npm:

```
npm install nodejs/i18n#module-experiment
```

### Usage: Raw Files

After you've run the `npm install` command above, you'll have all the localized markdown files from this repo accessible in your project at `node_modules/node-i18n/content/**/*.md`. (Note that the directory is called `node-i18n` here because that's the current `name` value in the package.json file. There's already a different module on npm with that name, so we may need to change it (cc @TiagoDanin).

I am new to Gatsby, but have been doing a bit of investigating. It seems like a Gatsby plugin like [gatsby-transformer-remark](https://www.gatsbyjs.org/packages/gatsby-transformer-remark/) or similar could be used to collect and read the contents of these files, and convert them from markdown to HTML.

```sh
# node majors
$ ls -A1 node_modules/node-i18n/content
v10.x
v12.x
v8.x
```

---

```sh
# locales
$ ls -A1 node_modules/node-i18n/content/v10.x
af-ZA
ar-SA
bg-BG
ca-ES
cs-CZ
da-DK
...
```

___

```sh
# markdown docs
$ tree -d node_modules/node-i18n/content/v10.x/en-US
└── doc
    ├── api
    ├── changelogs
    └── guides
        └── contributing
```

## Usage: API

In this PR I've also added a basic JavaScript API. To use the module:

```js
const i18n = require('node-i18n')
```

The module currently exports an object with the following properties:

- `allPages` - an Array of page objects that look like this:
  ```
  locale: 'af-ZA',
  nodeVersion: 'v10.x',
  filePath: 'doc/STYLE_GUIDE.md',
  fullPath: '/Users/z/git/nodejs/i18n/content/v10.x/af-ZA/doc/STYLE_GUIDE.md'
  ```
- `getPages` - a convenience method for getting a subset of the docs by `locale` and `nodeVersion`
- `locales` - a list of all the locale codes
- `nodeVersion` - a copy of the same object defined in package.json

### Feedback / Ideas?

I'd love people's thoughts on whether this is headed in the right direction. I am open to (and hoping for) other approaches than what I've laid out here.

cc @nodejs/website-redesign 
cc @nodejs/website 
cc @nodejs/i18n

alexandrtovmach commented:
Thanks for starting this conversation. I'm actually thinking that it's a bit overhead to create npm module for localization purposes. It makes sense to create any of npm package to share functionality (or content in our case) between many projects. We have just two nodejs.org and nodejs.dev and sounds like a good catch, but... in new website we're using Gatsby/GraphQL, that can be extended with [gatsby-source-git](https://www.gatsbyjs.org/packages/gatsby-source-git/) plugin to pull any of data directly from GitHub repo.

From my point of view we solved content delivery for nodejs.dev, so there no more reasons to create npm package, because it's going to be used just in one project - nodejs.org.

zeke commented:
From https://www.gatsbyjs.org/packages/gatsby-source-git 👍 

> Source plugin for pulling files into the Gatsby graph from abitrary Git repositories (hosted anywhere). This is useful if the markdown files you wish to render can’t live within your gatsby codebase, or if need to aggregate content from disparate repositories.

That's pretty cool I hadn't seen it. So is there work already underway in the nodejs.dev project to use that plugin?

alexandrtovmach commented:
@zeke no it's not in progress now, but it's going to be

benhalverson commented:
Does this module support `import i18n from 'node-i18n'` since we are in react land that is the prefered way over commonjs.

Can you include frontmatter for the docs?
An example from the learn docs. This is useful / might be required for the graphql queries.
It's also helpful for the Algolia search I've been working on.
```
---
title: Error handling in Node.js
description: 'How to handle errors during the execution of a Node.js application'
authors: flaviocopes, MylesBorins, fhemberger, LaRuaNa, ahmadawais
section: Getting Started
---
```

alexandrtovmach commented:
@benhalverson I think we're not going with this approach in nodejs.dev, and instead we'll use https://www.gatsbyjs.org/packages/gatsby-source-git

LaRuaNa commented:
> @benhalverson I think we're not going with this approach in nodejs.dev, and instead we'll use https://www.gatsbyjs.org/packages/gatsby-source-git

I'd also say let's pull it directly from github that saves us the overhead for package maintenance.

LaRuaNa commented:
#246

zeke commented:
We talked about getting this ready to ship in [today's meeting](https://github.com/nodejs/i18n/issues/307). I'm converting this back into a draft and have a few things to do to wrap this up:

- [ ] Add some usage documentation to the README
- [ ] Fix tests that have broken because of recent changes to `nodeVersions` structure in package.json

bnb commented:
cc @mhdawson this answers your question from yesterday's CommComm meeting ❤️

zeke commented:
I've updated the tests and added some documentation for the npm package. This is ready for review!

> This answers your question from yesterday's CommComm meeting

What was the question? :)

alexandrtovmach commented on guides/NPM_PACKAGE.md:
````diff
@@ -0,0 +1,89 @@
+You can install from the repo's master branch:
+
+```sh
+npm install nodejs/i18n
````
```suggestion
npm install nodejs/i18n
# or
yarn add nodejs/i18n
```

alexandrtovmach commented on guides/NPM_PACKAGE.md:
````diff
@@ -0,0 +1,89 @@
+Or install from a specific branch or tag:
+
+```sh
+npm install nodejs/i18n#some-branch
````
```suggestion
npm install nodejs/i18n#some-branch
# or
yarn add nodejs/i18n#some-branch
```

alexandrtovmach commented on guides/NPM_PACKAGE.md:
```diff
@@ -0,0 +1,89 @@
+v10.x
+v12.x
+v13.x
+v8.x
```
```suggestion
```
v8.x not supported anymore #263

alexandrtovmach commented on guides/NPM_PACKAGE.md:
```diff
@@ -0,0 +1,89 @@
+
+An array of all the locale codes used in the project, like `en-US`, `es-ES`, etc.
+
+### `nodeVersions`
```
```suggestion
### `supportedVersions`
```

alexandrtovmach commented on guides/NPM_PACKAGE.md:
```diff
@@ -0,0 +1,89 @@
+const i18n = require('node-i18n')
+
+// or..
+const { allPages, getPages, locales, nodeVersions } = require('node-i18n')
```
```suggestion
const { allPages, getPages, locales, supportedVersions } = require('node-i18n')
```

alexandrtovmach commented on guides/NPM_PACKAGE.md:
```diff
@@ -0,0 +1,89 @@
+# npm package
+
+This repository can be installed as an npm package. The goal is to make all of the Node.js project's English and localized content more easily consumable and re-usable in different projects.
```
```suggestion
This repository can be installed as an npm package. The goal is to make the English and localized Node.js docs more easily consumable and re-usable in different projects.
```

because, this module provides only docs, not a content for websites

JCMais commented on index.js:
```diff
@@ -0,0 +1,44 @@
+    return { locale, nodeVersion, filePath, fullPath }
+  })
+
+async function getPages (nodeMajor, locale) {
```
what is the reasoning behind `getPages` being async? Just to be future proof?

alexandrtovmach replied:
I think, better to declare it as `async` from a start, than later, when we should make some asynchronous stuff inside to not break backward compatibility

zeke replied:
Yep that was my thinking.

zeke commented:
Thanks for the suggestions. Committed them all. 👍
